---
title: "Assesing different normalization methods for spike-in RNA-seq"
author: "Przemyslaw Stempor"
date: "July 30, 2015"
output:
  ioslides_presentation:
    fig_retina: null
    smaller: yes
    transition: 0
---

<style>
.title-slide {
  background-color: #CBE7A5; /* #EDE0CF; ; #CA9F9D*/
}

<style>
.title-slide {
  background-color: #CBE7A5; /* #EDE0CF; ; #CA9F9D*/
}

slide {
  background-color: #FFFFFF; /* #EDE0CF; ; #CA9F9D*/
}


</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=TRUE)
```

## Data aquisition

- FASTQ files are aligned to ce10 reference genome using BWA
- Tags are summarized using coding gene model (only exonic reads) and ERCC spike-ins
- SummmarizedExperiment structures with annotations are acquired from JADB

```{r aquire, message=FALSE}
ContactExpIDs <- paste0("rFB0", 26:32)
require(JADBtools)
require(GenomicAlignments)
se <- getSummarizedEperiment(ContactExpIDs) 
rownames(se)[rownames(se) == ''] <- 
    as.character(elementMetadata(se)$geneName[rownames(se) == ''])
colnames(se) <- sapply(strsplit(colnames(se), '_'), '[[', 10)
show(se)
```


## Get spike-ins annotations
```{r aquireIN, message=FALSE}
tab <- read.table(
    system.file('spikeIN/cms_095046.txt', package = 'JADBtools'), 
    sep = '\t', header = TRUE, row.names = 2
)
attomole <- 602200
infoIN <- unlist(getDBdataField(paste0("rFB0", 26:32), field = 'SpikeIN'))
str(tab)
infoIN
```

## Construct DESeqDataSet and plot PCA
```{r, message=FALSE, dev='svg', fig.align='center', out.height=350, out.width="auto", fig.height=5.4, fig.width=9, dev.args=list(bg = "transparent")}
require(DESeq2); require(magrittr); require(ggplot2)
dds <- DESeqDataSet(se, ~strain)
plotPCA( 
    DESeqTransform( se ), intgroup = c("strain", "stage"), 
    ntop=Inf, returnData = TRUE 
)  %>% ggplot(., aes(PC1, PC2, col=strain, label=name, shape=stage))  + 
    geom_text(vjust=1.2) + geom_point(size=3) -> p; print(p)
```


## Calculate RPKM and gater data in plotable format
```{r, message=FALSE}
require(reshape2); require(dplyr);
rpkm <- fpkm(dds, robust = FALSE)
count <- melt(
    rpkm[grep('ERCC', rownames(rpkm)),], 
    varnames = c('spike','smpl'), value.name = 'count'
)

tab <- tab[grep('ERCC', rownames(rpkm), value = TRUE),]
cont <- melt( 
    tab[,grepl('2', infoIN)+3]*attomole,
    varnames = c('mix', 'spike'),value.name = 'cont'
)
data <- cbind( count, cont )
data  %>% mutate(cont=log2(cont), count=log2(count))  %>% 
    filter(!is.infinite(count)) -> data_log
head(data_log, 3)
```

## Plot spike-in regressions
```{r plotSpikeIN_func, include=FALSE}
plotSpikeIN <- function(dd, xpos = 23, ypos=-10) {
    library(proto)
    stat_smooth_func <- function (mapping = NULL, data = NULL, geom = "smooth", position = "identity",
                                  method = "auto", formula = y ~ x, se = TRUE, n = 80, fullrange = FALSE,
                                  level = 0.95, na.rm = FALSE, xpos=NULL, ypos=NULL, ...) {
        StatSmoothFunc$new(mapping = mapping, data = data, geom = geom, position = position,
                           method = method, formula = formula, se = se, n = n, fullrange = fullrange,
                           level = level, na.rm = na.rm, xpos=xpos, ypos=ypos, ...)
    }
    
    StatSmoothFunc <- proto(ggplot2:::Stat, {
        objname <- "smooth"
        
        calculate_groups <- function(., data, scales, method="auto", formula=y~x, ...) {
            rows <- daply(data, .(group), function(df) length(unique(df$x)))
            
            if (all(rows == 1) && length(rows) > 1) {
                message("geom_smooth: Only one unique x value each group.",
                        "Maybe you want aes(group = 1)?")
                return(data.frame())
            }
            
            # Figure out what type of smoothing to do: loess for small datasets,
            # gam with a cubic regression basis for large data
            # This is based on the size of the _largest_ group.
            if (identical(method, "auto")) {
                groups <- count(data, "group")
                
                if (max(groups$freq) < 1000) {
                    method <- "loess"
                    message('geom_smooth: method="auto" and size of largest group is <1000,',
                            ' so using loess.',
                            ' Use \'method = x\' to change the smoothing method.')
                } else {
                    method <- "gam"
                    formula <- y ~ s(x, bs = "cs")
                    message('geom_smooth: method="auto" and size of largest group is >=1000,',
                            ' so using gam with formula: y ~ s(x, bs = "cs").',
                            ' Use \'method = x\' to change the smoothing method.')
                }
            }
            if (identical(method, "gam")) try_require("mgcv")
            
            .super$calculate_groups(., data, scales, method = method, formula = formula, ...)
        }
        
        calculate <- function(., data, scales, method="auto", formula=y~x, se = TRUE, n=80, fullrange=FALSE, xseq = NULL, level=0.95, na.rm = FALSE, xpos=xpos, ypos=ypos, ...) {
            data <- remove_missing(data, na.rm, c("x", "y"), name="stat_smooth")
            if (length(unique(data$x)) < 2) {
                # Not enough data to perform fit
                return(data.frame())
            }
            
            if (is.null(data$weight)) data$weight <- 1
            
            if (is.null(xseq)) {
                if (is.integer(data$x)) {
                    if (fullrange) {
                        xseq <- scale_dimension(scales$x, c(0, 0))
                    } else {
                        xseq <- sort(unique(data$x))
                    }
                } else {
                    if (fullrange) {
                        range <- scale_dimension(scales$x, c(0, 0))
                    } else {
                        range <- range(data$x, na.rm=TRUE)
                    }
                    xseq <- seq(range[1], range[2], length=n)
                }
            }
            if (is.character(method)) method <- match.fun(method)
            
            method.special <- function(...)
                method(formula, data=data, weights=weight, ...)
            model <- safe.call(method.special, list(...), names(formals(method)))
            
            predictdf(model, xseq, se, level)
            m = model
            eq <- substitute(italic(y) == a + b %.% italic(x)*","~~italic(r)^2~"="~r2, 
                             list(a = format(coef(m)[1], digits = 3), 
                                  b = format(coef(m)[2], digits = 3), 
                                  r2 = format(summary(m)$r.squared, digits = 3)))
            func_string = as.character(as.expression(eq))
            
            data.frame(
                x=if(is.null(xpos)) min(data$x)*0.9 else xpos, 
                y=if(is.null(ypos)) max(data$y)*0.9 else ypos, 
                label=func_string
            )
        }
        
        required_aes <- c("x", "y")
        default_geom <- function(.) GeomSmooth
    })
    p <-  ggplot(dd, aes(cont, count, colour = smpl)) 
    p <- p + geom_point() 
    p <- p + geom_smooth(method=lm) 
    p <- p + facet_wrap(~smpl, nrow=2, scales="free")
    p <- p + xlab("RNA spike-in concentartion [RNA particles/ul]")
    p <- p + ylab("RNA expression measurement [RPKM]")
    #p + stat_smooth_func(geom="text",method="lm",hjust=-0.15,parse=TRUE, colour='black', size=4, vjust=11.5) 
    p <- p + facet_wrap(~smpl, nrow=2) 
    P <- p + stat_smooth_func(geom="text",method="lm",parse=TRUE, colour='black', size=4, xpos = xpos, ypos=ypos) 
    return(P)
}
```
```{r, message=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=8, fig.width=14, out.height=460, out.width="auto"}
plotSpikeIN(data_log)
```

## Plot spike-in regressions (RPKM>1)
```{r, message=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=8, fig.width=14, out.height=460, out.width="auto"}
data_log  %>% filter(count > 0)  %>% plotSpikeIN(xpos = 30, ypos = 0)
```

## Calculate regression coefficients and p-values (from F statistics)
```{r, message=FALSE}
library(plyr)
data_log  %>% filter(count > 0)  %>% ddply(.(smpl), function(df) {
    m <- lm(count ~ cont, data=df)
    data.frame(a = coef(m)[1], b = coef(m)[2], pval=anova(m)$'Pr(>F)'[[1]] )
}) -> coefs
coefs
```


## Normalizing for library depth, part 1

- Using all reads
```{r}
counts(dds, normalized = FALSE, replaced = FALSE)  %>% colSums -> read_sums
read_sums/mean(read_sums)
```

- Using spike-in reads
```{r}
counts(dds, normalized = FALSE, replaced = FALSE)[grep('ERCC', rownames(dds)),]  %>% colSums -> spike_sums
spike_sums/mean(spike_sums)
```

## Normalizing for library depth, part 2

- using the "median ratio method" described by Equation 5 in Anders and Huber (2010) for all reads
```{r, message=FALSE}
dds %>% estimateSizeFactors  %>% sizeFactors -> read_sf
read_sf/mean(read_sf)
```

- using the "median ratio method" described by Equation 5 in Anders and Huber (2010) for spike-in reads
```{r, message=FALSE}
dds[grep('ERCC', rownames(dds)),] %>% estimateSizeFactors  %>% 
    sizeFactors -> spike_sf
spike_sf/mean(spike_sf)
```

## Plot as barplot (normalized to 1)
```{r, message=FALSE}
dat <- data.frame(read_sums, spike_sums, read_sf, spike_sf)
barplot( 
    t(as.matrix(dat))/colMeans(as.matrix(dat)), 
    beside=TRUE, legend.text=names(dat) 
)
```

## Plot as barplot (normalized to 1)
```{r, message=FALSE}
dat <- data.frame(read_sums, spike_sums, read_sf, spike_sf)
barplot( 
    t(as.matrix(dat))/colMeans(as.matrix(dat)), 
    beside=TRUE, legend.text=names(dat) 
)
```

## Distribution of data in sampls
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=8, fig.width=14, out.height=450, out.width="auto"}
rpkm  %>% melt(value.name = 'x', varnames = c('gene', 'smpl'))  %>% 
    ggplot(aes(smpl, log2(x), col=smpl)) +  geom_boxplot(alpha=0.7)  + 
    geom_violin(alpha=0.5) + ylab('Expression [log2(RPKM)]') + xlab('Experiment')
```

## Distribution of spike-ins - plot 1
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=6, fig.width=14, out.height=350, out.width="auto"}
spikes <- grep("^ERCC", rownames(rpkm), value = TRUE)
infoIN  %>%  sapply(rep, spikes %>% length)  %>% melt(value.name = 'Mix')  %>% 
    dplyr::select(Mix)  %>% 
    cbind(melt(rpkm[spikes,]), subgroup=tab[spikes,]$subgroup, .)  %>% 
    ggplot(aes(x=Var1, y=log2(value), col=subgroup, shape=Mix)) + 
    geom_point(size=4) + scale_shape_manual(values = c(20, 2)) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

## Distribution of spike-ins - plot 2
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=6, fig.width=14, out.height=350, out.width="auto"}
infoIN  %>%  sapply(rep, spikes %>% length)  %>% melt(value.name = 'Mix')  %>% 
    dplyr::select(Mix)  %>% cbind(melt(rpkm[spikes,]), 
    subgroup=tab[spikes,]$subgroup, ., rat=tab[spikes,]$log2.Mix.1.Mix.2.)  %>% 
    ggplot(aes(x=rat, y=log2(value), col=subgroup, shape=Mix)) + 
    geom_violin(alpha=0.5) + geom_boxplot(alpha=0.5) +
    geom_jitter(size=4) + scale_shape_manual(values = c(20, 2))
```


## Colecting data for RUV normalization
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=6, fig.width=14, out.height=350, out.width="auto"}
zfGenes <- assays(se)$counts
zfGenes[spikes,infoIN=="Spike-in 2"] <- 
    zfGenes[spikes,infoIN=="Spike-in 2"]*tab[spikes,]$expected.fold.change.ratio
filter <- apply(zfGenes, 1, function(x) length(x[x>5])>=2)
filtered <- zfGenes[filter,]
genes <- rownames(filtered)[grep("^WB", rownames(filtered))]
spikes <- rownames(filtered)[grep("^ERCC", rownames(filtered))]
f <- factor(infoIN); levels(f) <- c('mix1', 'mix2');
colData(se)$mix <- f
colData(se)$spike <- colSums(zfGenes[spikes,])
colData(se)$sf <- zfGenes  %>% estimateSizeFactorsForMatrix
require(Heatplus)
```

## Spike-in expression w/o mix scaleng
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=6, fig.width=6, out.height=450, out.width="auto"}
plot(annHeatmap(log2(assays(se)$counts[spikes,]+1), ann=colData(se), cluster=list(cuth=10)))
```

## Spike-in expression witch mix scaleng
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=6, fig.width=6, out.height=450, out.width="auto"}
plot(annHeatmap(log2(zfGenes[spikes,]+1), ann=colData(se), cluster=list(cuth=10)))
```

## Define PCA plotting fun
```{r}
myPCA <- function(x) {
    x %>% DESeqTransform  %>% 
    plotPCA(intgroup = c("strain", "stage"), ntop=Inf, returnData = TRUE )  %>% 
    ggplot(., aes(PC1, PC2, col=strain, label=name, shape=stage))  + 
    geom_text(vjust=1.2) + geom_point(size=3)
}

```

## PCA plot, unnormalized data
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
se[genes]  %>% myPCA
```

## PCA plot, VST
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
se[genes]  %>% DESeqDataSet(~strain)  %>% DESeq -> ddsN
ddsN %>% varianceStabilizingTransformation %>% myPCA
```


## PCA plot, normalized counts (DEseq2)
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
log2(counts(ddsN, normalized=TRUE) + 1) %>% 
    SummarizedExperiment(colData=colData(ddsN))  %>% myPCA
```

## RUVg normalization using spike-in only
```{r RUV}
library(RUVSeq); library(EDASeq); library(RColorBrewer)
colData(se)$mix=NULL; colData(se)$sf=NULL
ruv <- RUVg(round(zfGenes), spikes, k=1)

differences <- matrix(data=c(c(1,2,4,5), c(3,6,7,-1)), byrow=TRUE, nrow=2)
ruvs <- RUVs(round(zfGenes), genes, k=1, differences)

require(edgeR)
design <- model.matrix(~strain, data=colData(se))
y <- DGEList(counts=assays(se)$counts[genes,], group=colData(se)$strain)
y <- calcNormFactors(y, method="upperquartile")
y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)
fit <- glmFit(y, design)
res <- residuals(fit, type="deviance")
ruvr <- RUVr(round(zfGenes), genes, k=1, res)

myHeatmap <- function(d, ...) {
    results(d) %>% .[order(.$padj),]  %>% head(50)  %>% rownames %>%
    counts(d)[.,]   %>% '+'(1)  %>% log2  %>% 
    annHeatmap(ann=colData(d), cluster=list(cuth=15, col=colors), ...)  %>% plot
}

colors <- brewer.pal(3, "Set2")
```

## Impact of RUVg
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=12, out.height=450, out.width="auto"}
par(mfrow=c(1,2))
plotRLE( zfGenes[genes,], col=colors[colData(se)[[1]]] )
plotRLE( ruv$normalizedCounts[genes,], col=colors[colData(se)[[1]]] )
```

## Impact of RUVg
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=12, out.height=450, out.width="auto"}
par(mfrow=c(1,2))
plotPCA( zfGenes[genes,], col=colors[colData(se)[[1]]] )
plotPCA( ruv$normalizedCounts[genes,], col=colors[colData(se)[[1]]] )
```

## PCA plot, RUVg and VSN
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
ruv$normalizedCounts[genes,]  %>% SummarizedExperiment(colData=colData(se))  %>%
    DESeqDataSet(~strain) %>% varianceStabilizingTransformation %>% myPCA
```


## Impact of RUVs
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=12, out.height=450, out.width="auto"}
par(mfrow=c(1,2))
plotRLE( zfGenes[genes,], col=colors[colData(se)[[1]]] )
plotRLE( ruvs$normalizedCounts[genes,], col=colors[colData(se)[[1]]] )
```

## Impact of RUVs
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=12, out.height=450, out.width="auto"}
par(mfrow=c(1,2))
plotPCA( zfGenes[genes,], col=colors[colData(se)[[1]]] )
plotPCA( ruvs$normalizedCounts[genes,], col=colors[colData(se)[[1]]] )
```

## PCA plot, RUVs and VSN
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
ruvs$normalizedCounts[genes,]  %>% SummarizedExperiment(colData=colData(se))  %>%
    DESeqDataSet(~strain) %>% varianceStabilizingTransformation %>% myPCA
```


## Impact of RUVr
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=12, out.height=450, out.width="auto"}
par(mfrow=c(1,2))
plotRLE( zfGenes[genes,], col=colors[colData(se)[[1]]] )
plotRLE( ruvr$normalizedCounts[genes,], col=colors[colData(se)[[1]]] )
```

## Impact of RUVr
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=12, out.height=450, out.width="auto"}
par(mfrow=c(1,2))
plotPCA( zfGenes[genes,], col=colors[colData(se)[[1]]] )
plotPCA( ruvr$normalizedCounts[genes,], col=colors[colData(se)[[1]]] )
```

## PCA plot, RUVr and VSN
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
ruvr$normalizedCounts[genes,]  %>% SummarizedExperiment(colData=colData(se))  %>%
    DESeqDataSet(~strain) %>% varianceStabilizingTransformation %>% myPCA
```


## Differential expression, raw data
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
zfGenes[genes,]  %>% DESeqDataSetFromMatrix(colData(se), ~strain)  %>% DESeq -> dds
dds  %>% myHeatmap
results(dds)  %>% .$padj  %>% '<'(0.05)  %>% table
```

## Differential expression, RUVg
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
ruv$normalizedCounts[genes,]  %>% DESeqDataSetFromMatrix(colData(se), ~strain)  %>% DESeq -> ddsG
ddsG %>% myHeatmap
results(ddsG)  %>% .$padj  %>% '<'(0.05)  %>% table
```

## Differential expression (edgeR), RUVg
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
design <- model.matrix(~strain + w, data=data.frame(colData(se), w=as.numeric(ruv$W)))
y <- DGEList(counts=assays(se)$counts[genes,], group=colData(se)$strain)
y <- calcNormFactors(y, method="upperquartile")
y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)
fit <- glmFit(y, design)
lrt <- glmLRT(fit, coef=2)
```

## Differential expression (edgeR), RUVg
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
topTags(lrt, 50) %>% rownames %>% ruv$normalizedCounts[.,]   %>% '+'(1)  %>% 
    log2  %>% annHeatmap(ann=colData(se), cluster=list(cuth=15, col=colors))  %>% plot
```

## Control: stage DE with raw counts
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
se[genes,]  %>% DESeqDataSet(~stage)  %>% DESeq %>% myHeatmap
```

## Control: stage DE with RUVg counts
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
ruv$normalizedCounts[genes,]  %>% DESeqDataSetFromMatrix(colData(se), ~stage)  %>% DESeq %>% myHeatmap
```

## Control: strain DE with RUVs counts
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
ruvs$normalizedCounts[genes,]  %>% DESeqDataSetFromMatrix(colData(se), ~strain)  %>% DESeq %>% myHeatmap
```

## Control: strain DE with RUVr counts
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
ruvr$normalizedCounts[genes,]  %>% DESeqDataSetFromMatrix(colData(se), ~strain)  %>% DESeq %>% myHeatmap
```


## Control: strain DE with SPIKE counts
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
ruvr$normalizedCounts[genes,]  %>% DESeqDataSetFromMatrix(colData(se), ~strain)  %>% DESeq %>% myHeatmap
```


## Control: spike counts vs. RUVr coefs
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
ruvr$normalizedCounts[genes,]  %>% DESeqDataSetFromMatrix(colData(se), ~strain)  %>% DESeq %>% myHeatmap
```

## Control: spike counts vs. RUVr coefs
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
data.frame(spike_sums, ruv$W, names(spike_sums))  %>% 
    ggplot(aes((spike_sums), W_1, label=names.spike_sums.)) + geom_text() + 
    geom_point() + geom_smooth(method=lm) + scale_y_reverse()
```

## Control: spike coefs vs. RUVr coefs
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
melt(
    assays(se)$counts[grep('ERCC', rownames(rpkm)),], 
    varnames = c('spike','smpl'), value.name = 'count'
)  %>% cbind( cont ) -> data_raw

data_raw %>% filter(count > 10)  %>% ddply(.(smpl), function(df) {
    m <- lm(cont ~ count, data=df)
    data.frame(a = coef(m)[1], b = coef(m)[2], pval=anova(m)$'Pr(>F)'[[1]] )
}) -> coef_raw
```

## Control: spike coefs vs. RUVr coefs
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=12, out.height=440, out.width="auto"}
plotSpikeIN(data_raw  %>% filter(count > 10), xpos =8.5, ypos=4) + 
    scale_x_log10() + scale_y_log10()
```

## Control: spike coefs vs. RUVr coefs
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
data.frame(spike_sums, coef_raw, ruv$W, names(spike_sums)) %>% 
     ggplot(aes(b/mean(b), W_1, label=names.spike_sums.)) + geom_text() + 
     geom_point() + geom_smooth(method=lm)
```

## Spike-in sums normalization
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
( filtered * (rep(1, nrow(filtered)) %*% t((spike_sums/1e3)/(read_sums/1e6))) )  %>%  
    "+"(1) %>% log2 %>% SummarizedExperiment(colData=colData(se))  %>% myPCA
```

## Spike-in sums normalization
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
( filtered * (rep(1, nrow(filtered)) %*% t((spike_sums/1e3)/(read_sums/1e6))) ) %>% 
    round %>% DESeqDataSetFromMatrix(colData(se), ~strain)  %>% DESeq %>% myHeatmap
```

## Linear model normalization
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
data_raw %>% filter(count > 10)  %>% ddply(.(smpl), function(df) {
    ids <- as.character(unique(df$smpl))
    m <- lm(cont ~ count, data=df)
    out <- predict(m, data.frame(count=as.integer(assays(se[genes,ids])$counts) ))
    return(out)
} ) %>% .[-1]  %>% as.matrix  %>% t -> norm_coef
#norm_coef <- round( norm_coef )
#colnames(norm_coef) <-  colnames(se)
```

## Spike-in Linear model
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
(norm_coef/attomole) %>% SummarizedExperiment(colData=colData(se))  %>% myPCA
```

## Spike-in Linear model
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
(norm_coef/attomole) %>% round %>% DESeqDataSetFromMatrix(colData(se), ~strain)  %>% DESeq %>% myHeatmap
```

## late embryo only 
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
LE <- se[,colData(se)$stage == 'LE']

filter <- apply(zfGenes[,colnames(LE)], 1, function(x) length(x[x>5])>=2)
filtered <- zfGenes[,colnames(LE)][filter,]
genes <- rownames(filtered)[grep("^WB", rownames(filtered))]
spikes <- rownames(filtered)[grep("^ERCC", rownames(filtered))]

ruv_le <- RUVg(round(zfGenes[,colnames(LE)]), spikes, 1)

differences <- rbind(c(1,2), c(3,4))
ruvs_le <- RUVs(round(zfGenes[genes,colnames(LE)] ), genes, k=1, differences)
```

## late embryo only RUVg
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
par(mfrow=c(1,2))
zfGenes[genes,colnames(LE)] %>% plotPCA(col=colors[colData(LE)[[1]]])
ruv_le$normalizedCounts[genes,]  %>% plotPCA(col=colors[colData(LE)[[1]]])
```

## late embryo only RUVs
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=12, out.height=450, out.width="auto"}
par(mfrow=c(1,2))
LE[genes,]  %>% assays  %>% .$counts  %>% plotPCA(col=colors[colData(LE)[[1]]])
ruvs_le$normalizedCounts[genes,]  %>% plotPCA(col=colors[colData(LE)[[1]]])
```

## late embryo only RUVs
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=12, out.height=450, out.width="auto"}
LE[genes,]  %>% DESeqDataSet(~strain)  %>% DESeq  %>% myHeatmap
```

## late embryo only RUVs
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
ruv_le$normalizedCounts[genes,] %>% DESeqDataSetFromMatrix( colData(LE), ~strain)  %>% DESeq  %>% myHeatmap
```

## late embryo only RUVs
```{r, message=FALSE, warning=FALSE, echo=TRUE, dev='svg', fig.align='center', fig.height=7, fig.width=8, out.height=450, out.width="auto"}
ruvs_le$normalizedCounts[is.na(as.integer(ruvs_le$normalizedCounts))] <- 0 
ruvs_le$normalizedCounts[genes,] %>% DESeqDataSetFromMatrix( colData(LE), ~strain)  %>% DESeq  %>% myHeatmap
```
